### PR TITLE
Fail earlier with invalid sink construction

### PR DESCRIPTION
**Pull request by osi**
"Fix" for #612 by moving failure to when the sink is constructed.

### CODE REVIEW

File: reactor-core/src/test/java/reactor/core/publisher/FluxProcessorTest.java [hunk 1]
```diff
@@ -261,4 +267,16 @@ public void scanProcessor() {
...
 		assertThat(test.scan(Scannable.BooleanAttr.TERMINATED)).isTrue();
 	}
+
+	@Test
+	public void shouldFailWhenUsingNonIgnoreStrategyOnInfiniteBuffer() throws Exception {
+		FluxProcessor<String, String> test = UnicastProcessor.create();
+
+		assertThat(test.getBufferSize()).isEqualTo(Integer.MAX_VALUE);
+
+		expectedException.expect(IllegalArgumentException.class);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxProcessorTest.java]: the way we now prefer to test exceptions is to use AssertJ:

```
Assertions.assertThatExceptionOfType(IllegalArgumentException.class)
    .isThrownBy(() -> test.sink(FluxSink.OverflowStrategy.BUFFER))
    .withMessage("Cannot use overflow strategy of BUFFER on UnicastProcessor with unlimited buffer");
```

File: reactor-core/src/main/java/reactor/core/publisher/FluxProcessor.java [hunk 2]
```diff
@@ -202,7 +201,9 @@ public Object scanUnsafe(Attr key) {
 	public final FluxSink<IN> sink(FluxSink.OverflowStrategy strategy) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxProcessor.java]: maybe hint at the limitation to IGNORE if buffer is unlimited, cite `TopicProcessor` as an example and say that using `onBackpressureXXX` on the processor afterward could still make sense?

Comment (osi) [inline reactor-core/src/main/java/reactor/core/publisher/FluxProcessor.java]: good idea. i'll enhance the javadocs too.

**Comment by smaldini** [comment]
To be investigated with #636 as I feel there's something to take away from this workaround behavior: the organization of the processors is not discoverable enough if we fallback to this.

**Comment by smaldini** [comment]
We'll revisit this PR with latest discussions on #612

**Pull request closed without merging**